### PR TITLE
[Payment method] Listing payment method

### PR DESCRIPTION
**Pull request by AdamKasp**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| License         | MIT

In Api we should have a unified way to get valid payment methods

### CODE REVIEW

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 1]
```diff
@@ -584,6 +605,30 @@ public function iShouldBeAbleToSelectPaymentMethod(string $paymentMethodName): v
         Assert::notFalse(array_search($paymentMethodName, array_column($paymentMethods, 'name'), true));
     }
 
+    /**
+     * @Then I should see :firstPaymentMethodName and :secondPaymentMethodName payment methods
+     */
+    public function iShouldSeeAndPaymentMethods(string $firstPaymentMethodName, string $secondPaymentMethodName): void
```
Comment (Roshyo) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:611]: I would name it `iShouldSeePaymentMethods` the plural form should be enough to give the hint of multiple ones. And the `And` here makes the function name hard to read

File: src/Sylius/Behat/Context/Setup/PaymentContext.php [hunk 2]
```diff
@@ -159,6 +160,22 @@ public function thePaymentMethodRequiresAuthorizationBeforeCapturing(PaymentMeth
         $this->paymentMethodManager->flush();
     }
 
+    /**
+     * @Given the store allows paying :paymentMethodName in :channel channel
+     */
+    public function theStoreAllowsPayingInChannel(string $paymentMethodName, ChannelInterface $channel): void
```
Comment (Roshyo) [inline src/Sylius/Behat/Context/Setup/PaymentContext.php:166]: `the store allows paying with` rather ?

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/OrderRepository.php [hunk 3]
```diff
@@ -431,4 +431,18 @@ public function findCartByTokenValue(string $tokenValue): ?BaseOrderInterface
...
+    {
+        return $this->createQueryBuilder('o')
+            ->andWhere('o.state = :state')
+            ->andWhere('o.tokenValue = :tokenValue')
+            ->andWhere('o.channel = :channel')
+            ->setParameter('state', OrderInterface::STATE_CART)
+            ->setParameter('tokenValue', $tokenValue)
+            ->setParameter('channel', $channel)
+            ->getQuery()
+            ->getOneOrNullResult()
```
Comment (Roshyo) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/OrderRepository.php:445]: Shouldn't we handle the `NonUniqueException` and return null ? Or do you think it is rather up to the calling script to do this ?
But in the second case, then we should add PHPDoc for the throw. Shouldn't we ?

File: UPGRADE-API-1.12.md [hunk 4]
```diff
@@ -0,0 +1,8 @@
+# UPGRADE FROM `v1.11.X` TO `v1.12.0`
+
+1. Service `src/Sylius/Bundle/ApiBundle/DataProvider/CartPaymentMethodsSubresourceDataProvider.php` has been removed and logic was replaced by `src/Sylius/Bundle/ApiBundle/DataProvider/PaymentMethodsCollectionDataProvider.php`
+
+2. Endpoints `/shop/orders/{tokenValue}/payments/{payments}/methods`, `/shop/payments/{id}/methods` has been removed and replaced by `/shop/payment-methods` with filter `paymentId` and `tokenValue`
+   `/shop/payment-methods` returns all enable payment methods if filters are not set, payment methods related to payment if filters are filled or empty response if filters ale filled with invalid data.
+
+
```
Comment (GSadee) [inline UPGRADE-API-1.12.md:8]: Too many blank lines

File: UPGRADE-API-1.12.md [hunk 5]
```diff
@@ -0,0 +1,8 @@
+# UPGRADE FROM `v1.11.X` TO `v1.12.0`
+
+1. Service `src/Sylius/Bundle/ApiBundle/DataProvider/CartPaymentMethodsSubresourceDataProvider.php` has been removed and logic was replaced by `src/Sylius/Bundle/ApiBundle/DataProvider/PaymentMethodsCollectionDataProvider.php`
```
Comment (GSadee) [inline UPGRADE-API-1.12.md:3]: Use namespaces instead of paths to identify the service

File: UPGRADE-API-1.12.md [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+
+2. Endpoints `/shop/orders/{tokenValue}/payments/{payments}/methods`, `/shop/payments/{id}/methods` has been removed and replaced by `/shop/payment-methods` with filter `paymentId` and `tokenValue`
```
Comment (GSadee) [inline UPGRADE-API-1.12.md:5]: Make this upgrade note consistent with changes in shipping methods after rebase :)

File: features/checkout/paying_for_order/viewing_available_payment_methods_based_on_current_channel.feature [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        And the store allows paying "Bank of Poland" in "Poland" channel
+        And the store allows paying offline for all channels
+        And the store allows paying "Bank of Universe"
+        And this payment method has been disabled
+        And the store has a product "PHP T-Shirt" priced at "$19.99" in "United States" channel
+        And this product is also priced at "$25.00" in "Poland" channel
+        And the store ships everywhere for free for all channels
+
+    @ui @api
+    Scenario: Seeing payment methods that are available in channel as an logged in customer
```
Comment (GSadee) [inline features/checkout/paying_for_order/viewing_available_payment_methods_based_on_current_channel.feature:23]:
```suggestion
    Scenario: Seeing payment methods that are available in channel as a logged in customer
```
here and in other scenarios

File: features/checkout/paying_for_order/viewing_available_payment_methods_based_on_current_channel.feature [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        Given I am a logged in customer
+        And I view shop on "United States" channel
```
Comment (GSadee) [inline features/checkout/paying_for_order/viewing_available_payment_methods_based_on_current_channel.feature:25]: This step looks like `Then` but it is used as `Given`, could you change it a little? For example to sth like:
```suggestion
        And I am in the "United States" channel
```

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php (hunk 1 above)
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:611]: You can use here `...` operator:
```suggestion
    public function iShouldSeeAndPaymentMethods(string ...$paymentMethodsNames): void
```

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 9]
```diff
@@ -584,6 +605,30 @@ public function iShouldBeAbleToSelectPaymentMethod(string $paymentMethodName): v
...
+    {
+        $paymentMethods = $this->getPossiblePaymentMethods();
+
+        Assert::inArray($firstPaymentMethodName, array_column($paymentMethods, 'name'));
+
+        Assert::inArray($secondPaymentMethodName, array_column($paymentMethods, 'name'));
+    }
+
+    /**
+     * @Then I should not see :firstPaymentMethodName and :secondPaymentMethodName payment methods
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:621]: Similar as above

File: src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutPaymentContext.php [hunk 10]
```diff
@@ -167,4 +167,36 @@ public function iShouldNotBeAbleToProceedCheckoutPaymentStep(): void
...
+        Assert::false(
+            $this->selectPaymentPage->hasPaymentMethod($firstPaymentMethodName),
+            sprintf('There is %s payment method', $firstPaymentMethodName)
+        );
+
+        Assert::false(
+            $this->selectPaymentPage->hasPaymentMethod($secondPaymentMethodName),
+            sprintf('There is %s payment method', $secondPaymentMethodName)
+        );
+    }
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutPaymentContext.php:201]: Refactor these methods to use `...` operator

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/OrderRepository.php (hunk 3 above)
Comment (AdamKasp) [reply src/Sylius/Bundle/CoreBundle/Doctrine/ORM/OrderRepository.php:445]: I think no ;) `tokenValue` is unique in DB so with the channel we are sure that we will get always a single cart or null :)
check this config: https://github.com/Sylius/Sylius/blob/master/src/Sylius/Bundle/CoreBundle/Resources/config/doctrine/model/Order.orm.xml#L76

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php (hunk 1 above)
Comment (AdamKasp) [reply src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:611]: it is an autogenerated name by behat be yea, in this case, I should change it :)

File: src/Sylius/Bundle/ApiBundle/DataProvider/PaymentMethodsCollectionDataProvider.php [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+
+        $parameters = $context['filters'];
+
+        if (!array_key_exists('tokenValue', $parameters)) {
+           return [];
+        }
+
+        if (!array_key_exists('paymentId', $parameters)) {
+           return [];
+        }
```
Comment (GSadee) [inline src/Sylius/Bundle/ApiBundle/DataProvider/PaymentMethodsCollectionDataProvider.php:58]: Still clear enough imo:
```suggestion
        if (!array_key_exists('tokenValue', $parameters) || !array_key_exists('paymentId', $parameters)) {
           return [];
        }
```

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/OrderRepository.php (hunk 3 above)
Comment (GSadee) [reply src/Sylius/Bundle/CoreBundle/Doctrine/ORM/OrderRepository.php:445]: `tokenValue` is unique, so IMO there is no need to use this exception

File: src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutPaymentContext.php [hunk 12]
```diff
@@ -167,4 +167,30 @@ public function iShouldNotBeAbleToProceedCheckoutPaymentStep(): void
...
+                $this->selectPaymentPage->hasPaymentMethod($paymentMethodName),
+                sprintf('There is no %s payment method', $paymentMethodName)
+            );
+        }
+    }
+
+    /**
+     * @Then I should not see :firstPaymentMethodName and :secondPaymentMethodName payment methods
+     */
+    public function iShouldNotSeeAndPaymentMethods(...$paymentMethodsNames): void
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutPaymentContext.php:187]:
```suggestion
    public function iShouldNotSeeAndPaymentMethods(string ...$paymentMethodsNames): void
```

File: src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutPaymentContext.php [hunk 13]
```diff
@@ -167,4 +167,30 @@ public function iShouldNotBeAbleToProceedCheckoutPaymentStep(): void
 
         throw new UnexpectedPageException('It should not be possible to complete checkout payment step.');
     }
+
+    /**
+     * @Then I should see :firstPaymentMethodName and :secondPaymentMethodName payment methods
+     */
+    public function iShouldSeeAndPaymentMethods(...$paymentMethodsNames): void
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Ui/Shop/Checkout/CheckoutPaymentContext.php:174]:
```suggestion
    public function iShouldSeeAndPaymentMethods(string ...$paymentMethodsNames): void
```

**Review by lchrusciel — APPROVED** [review]

**Comment by lchrusciel** [comment]
Thank you, Adam! :1st_place_medal:

File: features/checkout/paying_for_order/viewing_available_payment_methods_based_on_current_channel.feature [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+        And I am in the "United States" channel
+        And I have product "PHP T-Shirt" in the cart
+        When I complete addressing step with "United States" based billing address
+        And I complete the shipping step with first shipping method
+        Then I should be on the checkout payment step
+        And I should see "Bank of America" and "Offline" payment methods
+        But I should not see "Bank of Poland" and "Bank of Universe" payment methods
+
+    @api @ui
+    Scenario: Seeing shipping methods that are available in another channel as an logged in customer
```
Comment (GSadee) [inline features/checkout/paying_for_order/viewing_available_payment_methods_based_on_current_channel.feature:34]: `shipping` -> `payment`

File: features/checkout/paying_for_order/viewing_available_payment_methods_based_on_current_channel.feature [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+        Given I am in the "United States" channel
+        And I have product "PHP T-Shirt" in the cart
+        When I complete addressing step with email "john@example.com" and "United States" based billing address
+        And I complete the shipping step with first shipping method
+        Then I should be on the checkout payment step
+        And I should see "Bank of America" and "Offline" payment methods
+        But I should not see "Bank of Poland" and "Bank of Universe" payment methods
+
+    @ui @api
+    Scenario: Seeing shipping methods that are available in another channel as a guest
```
Comment (GSadee) [inline features/checkout/paying_for_order/viewing_available_payment_methods_based_on_current_channel.feature:55]: `shipping` -> `payment`

**Pull request merged**